include/seqan3/io/alignment_file/format_sam.hpp

```diff
@@ -0,0 +1,571 @@
+                          "The qual object must be a std::ranges::ForwardRange "
+                          "over letters that model seqan3::quality_concept.");
+
+        if constexpr (!detail::decays_to_ignore_v<mate_type>)
+            static_assert(detail::is_type_specialisation_of_v<remove_cvref_t<mate_type>, std::tuple>,
+                          "The mate object must be a std::tuple of size 3 with "
+                          "1) a std::ranges::ForwardRange with a value_type modelling seqan3::alphabet_concept, "
+                          "2) an std::UnsignedIntegral, and"
+                          "3) an std::UnsignedIntegral.");
+
+        if constexpr (!detail::decays_to_ignore_v<mate_type>)
+            static_assert((std::ranges::ForwardRange<decltype(std::get<0>(mate))>            &&
+                          alphabet_concept<value_type_t<decltype(std::get<0>(mate))>>        &&
+                          std::UnsignedIntegral<remove_cvref_t<decltype(std::get<1>(mate))>> &&
+                          std::UnsignedIntegral<remove_cvref_t<decltype(std::get<2>(mate))>>),
+                          "The mate object must be a std::tuple of size 3 with "
+                          "1) a std::ranges::ForwardRange with a value_type modelling seqan3::alphabet_concept, "
+                          "2) an std::UnsignedIntegral, and"
+                          "3) an std::UnsignedIntegral.");
+
+        if constexpr (!detail::decays_to_ignore_v<tag_dict_type>)
+            static_assert(std::Same<remove_cvref_t<tag_dict_type>, sam_tag_dictionary>,
+                          "The tag_dict object must be of type seqan3::sam_tag_dictionary.");
+
+        // ---------------------------------------------------------------------
+        // logical Requirements
+        // ---------------------------------------------------------------------
+        static_assert(detail::decays_to_ignore_v<align_type> ||
+                      (!detail::decays_to_ignore_v<seq_type> && !detail::decays_to_ignore_v<offset_type>),
+                      "If you specify an align object you must also specify the seq and offset object.");
+
+        // ---------------------------------------------------------------------
+        // Writing
+        // ---------------------------------------------------------------------
+        if constexpr (!detail::decays_to_ignore_v<header_type>)
+        {
+            if constexpr(!detail::decays_to_ignore_v<ref_id_type>)
+            {
+                auto ref = header.ref_dict | view::get<0>
+                                           | ranges::view::remove_if([&ref_id] (std::string & s) {return s != ref_id;});
+                if (ref.begin() == ref.end()) // no reference id matched
+                    throw format_error(std::string("The ref_id '") + ref_id + "' was not in the list of references");
+            }
+
+            if(!written_header)
+            {
+                check_header_information(header);
```

Comment:
The alignment_file_header object is quite unrestrictive and general as it takes mostly vectors and strings instead of a bloated enum/class/struct interface for simplicity. If we want to enforce that only valid SAM files are written with SeqAn we need to check the header object for correctness. Since this involves many things, I didn't do it yet for time reasons and just in case you have a different opinion on the checks.

Reply:
I would implement the checks inside the `write_header()` function and throw exceptions there.